python/ray/autoscaler/_private/kubernetes_operator/node_provider.py

```diff
@@ -0,0 +1,297 @@
+            extension = self.generate_extension(count, pod_spec, tags, new_node_names)
+            extensions.append(extension)
+            spec = self.generate_cluster_meta(body, pod_spec)
+            spec['extensions'] = extensions
+            body['spec'] = spec
+            logger.info(log_prefix + "calling create_namespaced_custom_object")
+            custom_object_api().create_namespaced_custom_object(
+                group=CRD_RAY_GROUP, version=CRD_RAY_VERSION,
+                namespace=self.namespace, plural=CRD_RAY_PLURAL, body=body, pretty='true')
+        else:
+            extensions = exist_cluster['spec']['extensions']
+            extensions.append(self.generate_extension(count, pod_spec, tags, new_node_names))
+            exist_cluster['spec']['extensions'] = extensions
+            body = exist_cluster
+            logger.info(log_prefix + "calling patch_namespaced_custom_object")
+            custom_object_api().patch_namespaced_custom_object(
+                group=CRD_RAY_GROUP, version=CRD_RAY_VERSION,
+                namespace=self.namespace, plural=CRD_RAY_PLURAL,
+                name=self.cluster_name, body=body)
+
+        new_svcs = []
+        if service_spec is not None:
+            logger.info(log_prefix + "calling create_namespaced_service "
+                        "(count={}).".format(count))
+
+            for new_node_name in new_node_names:
+
+                metadata = service_spec.get("metadata", {})
+                metadata["name"] = new_node_name
+                service_spec["metadata"] = metadata
+                service_spec["spec"]["selector"] = {"ray-node-uuid": node_uuid}
+                svc = core_api().create_namespaced_service(
+                    self.namespace, service_spec)
+                new_svcs.append(svc)
+
+        if ingress_spec is not None:
+            logger.info(log_prefix + "calling create_namespaced_ingress "
+                        "(count={}).".format(count))
+            for new_svc in new_svcs:
+                metadata = ingress_spec.get("metadata", {})
+                metadata["name"] = new_svc.metadata.name
+                ingress_spec["metadata"] = metadata
+                ingress_spec = _add_service_name_to_service_port(
+                    ingress_spec, new_svc.metadata.name)
+                extensions_beta_api().create_namespaced_ingress(
+                    self.namespace, ingress_spec)
+
+    def generate_cluster_meta(self, body, pod_spec):
+        spec = {}
+        body['apiVersion'] = CRD_RAY_GROUP + '/' + CRD_RAY_VERSION
+        body['kind'] = CRD_RAY_KIND
+        body['metadata'] = {}
+        body['metadata']['name'] = self.cluster_name
+        spec['clusterName'] = self.cluster_name
+        spec['images'] = {}
+        spec['images']['defaultImage'] = pod_spec['spec']['containers'][0]['image']
+        spec['imagePullPolicy'] = 'Always'
+        return spec
+
+    def generate_extension(self, count, pod_spec, tags, new_node_names):
```

Comment:
`generate_extension` is kind of fuzzy, can we use a more specific method name?

Reply:
Updated the function name.